taskcat/cfn/template.py

```diff
@@ -0,0 +1,133 @@
+import cfnlint
+from pathlib import Path
+from taskcat.client_factory import ClientFactory
+from taskcat.common_utils import s3_url_maker
+from typing import List, Set
+import logging
+import random
+import string
+
+LOG = logging.getLogger(__name__)
+
+
+class Template:
+
+    def __init__(self, template_path: str, project_root: str = '', url: str = '',
+                 client_factory_instance: ClientFactory = ClientFactory()):
+        self.template_path: Path = Path(template_path).absolute()
+        self.template = cfnlint.decode.cfn_yaml.load(self.template_path)
+        with open(template_path, 'r') as fh:
+            self.raw_template = fh.read()
+        project_root = project_root if project_root else self.template_path.parent.parent
+        self.project_root = Path(project_root).absolute()
+        self.client_factory_instance = client_factory_instance
+        self.url = url
+        self.children: List[Template] = []
+        self._find_children()
+
+    def __str__(self):
+        return str(self.template)
+
+    def __repr__(self):
+        return "<Template {} at {}>".format(self.template_path, hex(id(self)))
+
+    def _upload(self, bucket_name: str, prefix: str = '') -> str:
+        s3_client = self.client_factory_instance.get('s3')
+        s3_client.upload_file(str(self.template_path), bucket_name, prefix + self.template_path.name)
+        return s3_url_maker(bucket_name, f'{prefix}{self.template_path.name}', self.client_factory_instance.get('s3'))
+
+    def _delete_s3_object(self, url):
+        if url:
+            bucket_name = url.split('//')[1].split('.')[0]
```

Comment:
I'm concerned we'll run into legacy sig-v2 URLs here.

Reply:
not sure if I understand where/why sigv2 would be used...

as a side note, using virtual host style urls `https://<BUCKET_NAME>.s3-<REGION>.amazonaws.com/<KEY>` so future proof from path style url deprecation

Reply:
tested with a bucket in a v4 only region (eu-north-1) and can confirm requests are using sigv4.